File: apps/api/src/config/flags.ts
/**
 * Boolean environment flags, in one place.
 *
 * `config.ts` parses them as part of the validated `ApiConfig`, and the tracer
 * bootstrap parses `OBSERVABILITY_ENABLED` on its own — it runs before the
 * configuration module can be imported, because loading that module would pull
 * in the very clients `dd-trace` has to patch first. Two readers, one definition
 * of what counts as true, so the switch cannot mean one thing to the tracer and
 * another to the rest of the service.
 *
 * This file deliberately imports nothing.
 */

export const TRUTHY_FLAG_VALUES: readonly string[] = ['1', 'true', 'yes', 'on'];
export const FALSY_FLAG_VALUES: readonly string[] = ['0', 'false', 'no', 'off'];

/** Human-readable list for a validation message. */
export const BOOLEAN_FLAG_HINT = `must be a boolean: one of ${[
  ...TRUTHY_FLAG_VALUES,
  ...FALSY_FLAG_VALUES,
]
  .filter((value, index, all) => all.indexOf(value) === index)
  .join(', ')}`;

/**
 * `true` or `false` for a recognized value, `null` for anything else — including
 * an absent or blank variable. Callers decide whether an unrecognized value is a
 * startup failure (configuration) or simply "off" (the tracer bootstrap, which
 * has no logger yet and must not crash the process over a typo it cannot report).
 */
export function parseBooleanFlag(value: string | undefined): boolean | null {
  if (value === undefined) {
    return null;
  }
  const normalized = value.trim().toLowerCase();
  if (TRUTHY_FLAG_VALUES.includes(normalized)) {
    return true;
  }
  if (FALSY_FLAG_VALUES.includes(normalized)) {
    return false;
  }
  return null;
}

File: apps/worker/src/publishhub_worker/config/flags.py
"""
Boolean environment flags, in one place.

Mirrors `apps/api/src/config/flags.ts`, value for value. `OBSERVABILITY_ENABLED`
is read by two things — the configuration loader and, later, the tracer bootstrap
that has to run before anything else is imported — so what counts as true is
defined once instead of twice.

This module deliberately imports nothing.
"""

from __future__ import annotations

TRUTHY_FLAG_VALUES: tuple[str, ...] = ("1", "true", "yes", "on")
FALSY_FLAG_VALUES: tuple[str, ...] = ("0", "false", "no", "off")

#: Human-readable list for a validation message.
BOOLEAN_FLAG_HINT = "must be a boolean: one of " + ", ".join(
    (*TRUTHY_FLAG_VALUES, *FALSY_FLAG_VALUES)
)


def parse_boolean_flag(value: str | None) -> bool | None:
    """
    `True` or `False` for a recognized value, `None` for anything else —
    including an absent or blank variable. The caller decides whether an
    unrecognized value is a startup failure (configuration, which can report the
    offending key) or simply "off" (a bootstrap with no logger yet, which must
    not crash the process over a typo it cannot report).
    """
    if value is None:
        return None
    normalized = value.strip().lower()
    if normalized in TRUTHY_FLAG_VALUES:
        return True
    if normalized in FALSY_FLAG_VALUES:
        return False
    return None
